Backport 1.3: Fix corner case uses of memory_buffer_alloc.c

andresag01 opened the pull request:
This change fixes the following corner cases in memory_buffer_alloc.c:
* Calling `verify_chain()` on an uninitialized chain. Would previously cause a `NULL` dereference.
* Allocating a memory block large enough to overflow the length variable.
* Initializing the allocator with a buffer smaller than the size of a memory header.

This PR is based on the information recorded in https://github.com/ARMmbed/mbedtls/issues/639.

**NOTES:**
* This PR includes tests.
* This is a backport for #778

gilles-peskine-arm commented:
Failure in `x509parse-suite` (https://jenkins-internal.mbed.com/job/mbedtls-pr/425/execution/node/192/log/). **DO NOT MERGE** pending investigation.

gilles-peskine-arm commented on library/x509_crl.c:
```diff
@@ -295,7 +295,7 @@ int x509_crl_parse_der( x509_crl *chain,
      * Copy raw DER-encoded CRL
      */
-    if( ( p = polarssl_malloc( buflen ) ) == NULL )
+    if( buflen != 0 && ( ( p = polarssl_malloc( buflen ) ) == NULL ) )
```
The check for `buflen == 0` should be made earlier. See https://github.com/ARMmbed/mbedtls/pull/1009/files#r153007520

andresag01 commented:
@gilles-peskine-arm: I have fixed the issue.

andresag01 commented:
Changed the code in x509_crl.c according to the latest review comments.

gilles-peskine-arm commented:
Automatic CI run picked an old version. Running manually: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/46/

andresag01 commented:
@gilles-peskine-arm: I found additional issues when running this branch on ASan. I will look at it and fix it as soon as I can.

andresag01 commented:
@gilles-peskine-arm: I have rebased the changes and added a few more fixes. Turns out that this version of memory_buffer_alloc.c had diverged a lot more from development and mbedtls-2.1 than I expected, and there were some memory leaks and missing checks in here compared to the more recent versions of the library.

sbutcher-arm commented:
Mbed TLS 1.3 is now End of Life, as has been [previously announced](https://tls.mbed.org/tech-updates/releases/mbedtls-2.7.0-2.1.10-and-1.3.22-released).

Unfortunately that means no further releases of Mbed TLS 1.3 will be made, so this PR will not be merged, and is closed now.
